File: 15/Part2.ts
console.log('Starting Async')
console.time('ASync')
import { Divisor, Factors, Start } from "./input";



let factorA = Factors.A
async function updateA(start: number) {
  let ret = start
  do {
    ret = (ret*factorA)%Divisor
  } while(ret%4)
  return ret
}
let factorB = Factors.B
async function updateB(start: number) {
  let ret = start
  do {
    ret = (ret*factorB)%Divisor
  } while(ret%8)
  return ret
}

async function compare({A, B}: {A: number, B: number}) {
  let [genA, genB] = await Promise.all([updateA(A), updateB(B)])
  let [a, b] = await Promise.all([trimAndPad(genA), trimAndPad(genB)])
  return {A: genA, B: genB, same: a===b}
}

async function trimAndPad(num: number) {
  let str = num.toString(2).slice(-16)
  while (str.length<16) str = '0'+str
  return str
}

async function run() {
  let pair = Start
  //let pair = {A:116, B:299}
    , count = 0
  for (let i = 0; i < 5000000; i++) {
    let current = await compare(pair)
    if(current.same) count++
    pair = current
  }
  return count
}

async function DOTHETHING() {
  let a = await run()
  console.log(a)
  console.timeEnd('ASync')
}

DOTHETHING()

File: 15/Part2.lua
local time = os.clock()
local band = bit32 and bit32.band or bit.band
local divisor = 2147483647
local factors = {
  A = 16807,
  B = 48271,
}
local start = {
  A = 679,
  B = 771
}

local current, count = start, 0

for i=1, 5000000 do
  repeat
    current.A = (current.A*factors.A)%divisor
  until((current.A%4)==0)
  repeat
    current.B = (current.B*factors.B)%divisor
  until((current.B%8)==0)

  if (band(current.A,0xFFFF) == band(current.B,0xFFFF)) then count = count + 1 end
end

print(count)
print(os.clock()-time)